api/segments/models.py

```diff
@@ -261,13 +300,44 @@ def get_segment(self):
             rule = rule.rule
         return rule.segment
 
+    def matches_rule(self, rule: "SegmentRule") -> bool:
+        if rule.type != self.type:
+            return False
+
+        conditions = rule.conditions.all()
+        _conditions = self.conditions.all()
+
+        if not conditions and not _conditions:
+            # Empty rule with the same type matches.
+            return True
+
+        matched_conditions = set()
+
+        for condition in conditions:
+            for _condition in _conditions:
+                if _condition in matched_conditions:
+                    continue
```

Comment:
I don't really understand the need for this check?

Reply:
Because this part here:

```python
for _condition in _conditions:
```

Is subnested in the parent 

```python
for condition in conditions:
```

So if we have already matched a `_condition` we do not want to evaluate it again.